libpromises/files_names.c

```diff
@@ -674,11 +674,29 @@ char *GetRealPath(const char *const path)
         return NULL;
     }
     char *const abs_path = GetAbsolutePath(path);
-    if (abs_path == NULL || abs_path[0] == '\0')
+    if (NULL_OR_EMPTY(abs_path))
     {
-        return abs_path;
+        free(abs_path);
+        return NULL;
+    }
+
+#ifdef __linux__ // POSIX 2008 - could add newer versions of BSD / solaris
+    char *real_path = realpath(abs_path, NULL);
+    if (NOT_NULL_AND_EMPTY(real_path))
+    {
+        free(real_path);
+        real_path = NULL;
     }
-    char *const real_path = realpath(abs_path, NULL);
+#else // Pre POSIX 2008 - realpath arg cannot be NULL
+    char *const path_buf = xcalloc(1, PATH_MAX);
+    char *real_path = realpath(abs_path, path_buf);
+    if (NULL_OR_EMPTY(real_path))
+    {
+        free(path_buf);
+        real_path = NULL;
+    }
+#endif
```

Comment:
Why don't we always pass a statically allocated buffer of size ``PATH_MAX``? To me that looks better than this ifdef-else-endif machinery.

Reply:
From realpath man page:

Reply:
> BUGS

> The POSIX.1-2001 standard version of this function is broken by design, since it is impossible to determine a suitable size for the output buffer, resolved_path.
> According to POSIX.1-2001 a buffer of size PATH_MAX suffices, but PATH_MAX need not be a defined constant, and may have to be obtained using pathconf(3).
> And asking pathconf(3) does not really help, since, on the one hand POSIX warns that the result of pathconf(3) may be huge and unsuitable for mallocing memory, and on the other hand pathconf(3) may return -1 to signify that PATH_MAX is not bounded.
> The resolved_path == NULL feature, not standardized in POSIX.1-2001, but standardized in POSIX.1-2008, allows this design problem to be avoided.

Reply:
Yes, but then you cannot rely on ``PATH_MAX`` at all.

Reply:
Well, I don't have a good alternative. I prefer the default (linux) to use the NULL argument which is more correct, and safer. We do set `PATH_MAX` to 4096 by default, so this way, it should still work using PATH_MAX on exotics. (Our code already relies on `PATH_MAX`, in many other places.)